src/Suggestor/Suggestor.ts

```diff
@@ -378,14 +407,86 @@ function addRecurrenceSuggestions(
     return results;
 }
 
+function addIDSuggestion(line: string, cursorPos: number, idSymbol: string, allTasks: Task[]) {
+    const results: SuggestInfo[] = [];
+    const idRegex = new RegExp(`(${idSymbol})\\s*([0-9a-zA-Z ]*)`, 'ug');
+    const idMatch = matchIfCursorInRegex(line, idRegex, cursorPos);
+
+    if (idMatch && idMatch[0].trim().length <= idSymbol.length) {
+        const ID = generateUniqueId(allTasks.map((task) => task.id));
+        results.push({
+            suggestionType: 'match',
+            displayText: 'Auto Generate Unique ID',
+            appendText: `${idSymbol} ${ID}`,
+            insertAt: idMatch.index,
+            insertSkip: idSymbol.length,
+        });
+    }
+
+    return results;
+}
+
+/*
+ * If the cursor is located in a section that is followed by a Depends On Symbol, suggest options
+ * for what to enter as Depend on Option.
+ * It should contain suggestion of Possible Dependant Tasks
+ * of what the user is typing.
+ */
+function addDependsOnSuggestions(
+    line: string,
+    cursorPos: number,
+    settings: Settings,
+    dependsOnSymbol: string,
+    allTasks: Task[],
+    taskToSuggestFor?: Task,
+) {
+    const results: SuggestInfo[] = [];
+
+    const dependsOnRegex = new RegExp(`(${dependsOnSymbol})([0-9a-zA-Z ^,]*,)*([0-9a-zA-Z ^,]*)`, 'ug');
```

Comment:
Hi @Genei180,

I'm writing tests for the new behaviour, trying to get good coverage of all the behaviours, before fixing one or two bugs I found in testing this feature in my own main vault.

I'm having a bit of trouble understanding the intention of some bits of the regular expression - can you remember the reasons for any of these:

- Why is `(${dependsOnSymbol})` captured as a group?
- Why is a space character included in `[0-9a-zA-Z ^,]`?
- Why are the number of occurrences all `*` rather than `+` - in other words, why is 0 occurrences allowed throughout?

I'm asking because I want to do some refactoring to make use of an existing regex... below ... to specify the allowed characters in IDs, so first I need to add tests for all the current behaviours of that regular expression - which means understanding its requirements...

```ts
// The allowed characters in a single task id:
export const taskIdRegex = /[a-zA-Z0-9-_]+/;
```

Thanks in advance....

Reply:
Hi @claremacrae, 

Will Happily Help.
A helpfull Resource can be [regex101](https://regex101.com/)

It Provides Easy Visualisation (Left), Explanation to the Right (Tab Closed, Right), and Displays the Group Matches(Right).
![image](https://github.com/obsidian-tasks-group/obsidian-tasks/assets/16257607/c62457fd-9347-410c-96ba-b4b510803e12)

I tried to make it as Verbose as Possible, with the Following Variable Assignments, too Explain what each Capture Group, does/Matches.

```
// dependsOnMatch[1] = Depends On Symbol
const existingDependsOnIdStrings = dependsOnMatch[2] || '';
const newTaskToAppend = dependsOnMatch[3];
```

>  Why is (${dependsOnSymbol}) captured as a group?

Good Question, and you're Right it could Probably be removed.
Used Originally, and then Replaced it with the Passed in Parameter.
So i forgot too Change the Regex, and it became redundant.

> Why is a space character included in [0-9a-zA-Z ^,]?

It Catches any Spaces, that were put there for Formatting Reasons.
The Second Capture Group, captures, all Existing ID's, to re-add to the Suggestion.
You could Probably, remove it, and just adjust the insert At Index, to append it.
But i found it Useful, to match the Whole Depends on as Regex.

If the Space is not in the Second Regex Capture Group, you can't Distinguish Existing IDs, form Typed Text:
![image](https://github.com/obsidian-tasks-group/obsidian-tasks/assets/16257607/84c1ac53-e577-4108-9aa3-a3e9355c0681)
All is matched, in Capture Group 3, and you need too Split it up anyway.

In Other Words, it does not Match ID's, but all via ", " Chained ID's, till the last ",".
Everything after is Interpreted as Text too Search for in Suggested Tasks.

>  Why are the number of occurrences all * rather than + - in other words, why is 0 occurrences allowed throughout?

Because, the Suggestion needs to occur, after the DependsOn Symbol, even if nothing is present.

> I'm asking because I want to do some refactoring to make use of an existing regex.

I guess the Motivation behind is readability. I understand that.
You could Probably Split up the Regex in Multiple Parts.
For my Part, i didn't find this more Readable or easier too Understand.
I tried solving that over the Variable Name, which explains what the Capture Group does.
This should guarantee short, easy too read Code.

If i could have written something more Clearly or better, please let me know so i can learn from it.

Additionally, I can assist you with testing, but I'm not Very Experienced what to Test for.
So I need to Input like, "Please, write Tests for the Regex Capture Group" or something similar, so I know where to start and what to Test for.

Reply:
Hi, thank you for a very helpful reply indeed....

I'm thinking about a mixture of understanding the behaviour, testing it, and maintaining it.

Here's a bit of a brain dump.

tl;dr 

1. the first thing to do is to fix the new test which is marked as `.failing` - see below...
2. sorry to be giving this feedback now rather than before merging... I normally try and do this level of testing before merging, but needed to have the new code in my vault to be able to test it with real data....
3. it might be easier if we had a video call about this rather than typing here - I have an email address in my GitHub profile if this is of interest to you...

### Allowed characters in IDs in Tasks

The valid characters in IDs are **documented** in [ID](https://publish.obsidian.md/tasks/Getting+Started/Task+Dependencies#%60id%60).

The valid characters in IDs are **defined** in a single source of truth:

https://github.com/obsidian-tasks-group/obsidian-tasks/blob/fd391f29c69231711188d83fc6eab53257dc066a/src/TaskSerializer/DefaultTaskSerializer.ts#L48-L49

`taskIdRegex` is then used in several other places in the code - so that if we ever change the rules, there is only one place to edit.

### Allowed characters in IDs in Auto Suggest

 When using the new feature in my own vault, I found that the new feature does not recognise `-` and `_` as valid characters in IDs.

This means that with the following markdown, I cannot use auto-suggest to make 'another task' depend on a second task, as it does not recognise `⛔ 12_34` as a valid ID:

```
- [ ] hello world 🆔 12_34
- [ ] another task ⛔ 12_34, 
```

We could obviously just add those two characters to the regexes in the new Auto Suggest code, although my preference would be to eventually refactor the new code so that all ID regexes reuse `taskIdRegex`.

### Understanding the regular expression I asked about

I previously put the new depends-on regular expression into regex101, with a selection of possible task lines in - sensible and not - to understand the behaviour of the regex: See https://regex101.com/r/0tPwyZ/1, and the screenshot below.

This is why I asked about the use of `*` as opposed to `+`.

I suspect that the regex may be too flexible, and the screenshot shows it is allowing users to construct 'depends-on' strings that Tasks would not actually parse (because it reads back from the end of the line until it finds something that is neither a recognised Tasks field nor a tag)

![image](https://github.com/obsidian-tasks-group/obsidian-tasks/assets/4840096/11149443-3371-40e1-b6d6-9d07ec4acb03)

### Adding tests

I've made some progress adding tests for the new code:

https://github.com/obsidian-tasks-group/obsidian-tasks/blob/fd391f29c69231711188d83fc6eab53257dc066a/tests/Suggestor/Suggestor.test.ts#L251-L325

There is one test with `.failing` - which is the one that uses `_` and `-` in an ID.

This would be the priority to fix, as it is definitely blocking release.

Reply:
This is the full PR where I added some tests...
https://github.com/obsidian-tasks-group/obsidian-tasks/pull/2806

Reply:
> There is one test with `.failing` - which is the one that uses `_` and `-` in an ID.
> 
> This would be the priority to fix, as it is definitely blocking release.

I'm working on this...